### PR TITLE
validation sum of replicas and parallelism inside int32 range

### LINKED ISSUE #169: Webhook validation for int32 physical type limitation
**What would you like to be added**:
Webhook validation for int32 physical type limitation.
**Why is this needed**:
Now when the ReplicatedJobStatus.Ready is calculated the maximum possible value it can store is not taken into account.
int32       the set of all signed 32-bit integers (-2147483648 to 2147483647)

/kind bug

### PR DESCRIPTION
**Pull request by stuton**
Fixes #169

/kind bug

### CODE REVIEW

File: api/jobset/v1alpha1/jobset_webhook.go [hunk 1]
```diff
@@ -73,6 +74,12 @@ func (js *JobSet) ValidateCreate() error {
 	var allErrs []error
 	// Validate that replicatedJobs listed in success policy are part of this JobSet.
 	validReplicatedJobs := replicatedJobNamesFromSpec(js)
+	for _, rjob := range js.Spec.ReplicatedJobs {
+		sum := int64(*(rjob.Template.Spec.Parallelism)) + int64(rjob.Replicas)
```
Comment (ahg-g) [inline api/jobset/v1alpha1/jobset_webhook.go:78]: A couple of things:
1) multiply not sum (replicas * parallelism < maxint)
2) check if rjob.Template.Spec.Parallelism is nil, in which case it parallelism is 1

File: api/jobset/v1alpha1/jobset_webhook.go [hunk 2]
```diff
@@ -73,6 +74,12 @@ func (js *JobSet) ValidateCreate() error {
...
+		if sum > math.MaxInt32 || sum < math.MinInt32 {
+			allErrs = append(allErrs, fmt.Errorf("the sum replicas and parallelism are out of range int32 type in replicatedJob name '%s'", rjob.Name))
```
Comment (ahg-g) [inline api/jobset/v1alpha1/jobset_webhook.go:80]:
```suggestion
			allErrs = append(allErrs, fmt.Errorf("the product of replicas and parallelism must not exceed %d for replicatedJob '%s'", math.MaxInt32, rjob.Name))
```

File: api/jobset/v1alpha1/jobset_webhook_test.go [hunk 3]
```diff
@@ -351,3 +355,69 @@ func TestJobSetDefaulting(t *testing.T) {
...
 }
+
+func TestValidateCreate(t *testing.T) {
+	testCases := []struct {
+		name string
+		js   *JobSet
+		want error
+	}{
+		{
+			name: "reach out of int32 range",
```
Comment (ahg-g) [inline api/jobset/v1alpha1/jobset_webhook_test.go:366]:
```suggestion
			name: "number of pods exceeds the limit",
```

File: api/jobset/v1alpha1/jobset_webhook_test.go [hunk 4]
```diff
@@ -351,3 +355,69 @@ func TestJobSetDefaulting(t *testing.T) {
...
+					SuccessPolicy: &SuccessPolicy{},
+				},
+			},
+			want: errors.Join(
+				fmt.Errorf("the sum replicas and parallelism are out of range int32 type in replicatedJob name 'test-jobset-replicated-job-0'"),
+				fmt.Errorf("the sum replicas and parallelism are out of range int32 type in replicatedJob name 'test-jobset-replicated-job-1'"),
+			),
+		},
+		{
+			name: "inside of int32",
```
Comment (ahg-g) [inline api/jobset/v1alpha1/jobset_webhook_test.go:398]:
```suggestion
			name: "number of pods within the limit",
```

File: api/jobset/v1alpha1/jobset_webhook.go (hunk 1 above)
Comment (stuton) [reply api/jobset/v1alpha1/jobset_webhook.go]: perhaps, you meant (replicas * parallelism > maxint)?

File: api/jobset/v1alpha2/jobset_webhook.go [hunk 5]
```diff
@@ -73,6 +74,14 @@ func (js *JobSet) ValidateCreate() error {
 	var allErrs []error
 	// Validate that replicatedJobs listed in success policy are part of this JobSet.
 	validReplicatedJobs := replicatedJobNamesFromSpec(js)
+	for _, rjob := range js.Spec.ReplicatedJobs {
+		if rjob.Template.Spec.Parallelism == nil {
+			rjob.Template.Spec.Parallelism = pointer.Int32(1)
```
Comment (ahg-g) [inline api/jobset/v1alpha2/jobset_webhook.go:79]: we are not supposed to change the value, just use a temp variable named `parallelism`

```
parallelism = 1
if rjob.Template.Spec.Parallelism != nil {
  parallelism = *rjob.Template.Spec.Parallelism
}
```

**Pull request merged**